server/src/controllers/PartyController.ts

```diff
@@ -0,0 +1,345 @@
+      address: userAddress
+    });
+    if (!grail) {
+      res.status(404).send({ type: "notFound", userAddress });
+      return;
+    }
+    const partyAddress = PartyController.trimAndToLower(req.params.address);
+    const result = await this.partyCollection.findOneAndUpdate(
+      {
+        address: partyAddress,
+        userlist: { $nin: [userAddress] },
+        pendingUserlist: { $nin: [userAddress] }
+      },
+      {
+        $addToSet: {
+          pendingUserlist: userAddress
+        }
+      },
+      { returnOriginal: false }
+    );
+    if (result && result.ok && result.value) {
+      res.json({ success: true });
+      return;
+    }
+    // We didnt get a result, so there must already be a user with this address in the specified party
+    res.status(409).send({ type: "conflict", userAddress });
+  };
+
+  public updatePartyUser = async (req: Request, res: Response) => {
+    try {
+      const method = req.body.method;
+      switch (method) {
+        case "accept":
+          this.acceptPartyUser(req, res);
+          break;
+        case "deny":
+          this.denyPartyUser(req, res);
+          break;
+        case "remove":
+          this.removePartyUser(req, res);
+          break;
+        default:
+          res.status(404).send({ type: "methodUnknown", method });
+      }
+    } catch (err) {
+      PartyController.sendUnknownError(res, err);
+    }
+  };
+
+  private acceptPartyUser = async (req: Request, res: Response) => {
+    const password = req.body.password;
+    const address = PartyController.trimAndToLower(req.body.address);
+    const token = req.body.token;
+    const acceptedUser = req.body.user;
+    const result = await this.partyCollection.findOne({
+      address: address,
+      password: password,
+      token: token
+    });
+    if (result) {
+      const newUserGrailData = await this.pullUserGrailData(acceptedUser);
+      var newUsers = [];
+      var newUserlist = [];
+      var foundCurrentEntry = false;
+      result.data.users.forEach(user => {
+        if (user.username === acceptedUser) {
+          newUsers.push(newUserGrailData);
+          foundCurrentEntry = true;
+        } else {
+          newUsers.push(user);
+        }
+        newUserlist.push(user.username);
+      });
+      if (!foundCurrentEntry) {
+        newUsers.push(newUserGrailData);
+        newUserlist.push(acceptedUser);
+      }
+      result.data.users = newUsers;
+      const userPendingIndex = result.pendingUserlist.indexOf(acceptedUser);
+      if (userPendingIndex !== -1) {
+        result.pendingUserlist.splice(userPendingIndex, 1);
+      }
+      this.update(res, address, password, token, dataToSet => {
+        dataToSet.data = result.data;
+        dataToSet.userlist = newUserlist;
+        dataToSet.pendingUserlist = result.pendingUserlist;
+        return dataToSet;
+      });
+      return;
+    }
+    // Didnt get the party document. Handle the error
```

Comment:
I think if you change the db model as mentioned above, you can probably remove all this logic and this can then be a simple call something like (pseudo code)

```
filter: {address: xyz, password: whatever},
userList: addToSet(acceptedUser)
pendingList: pullFromSet(acceptedUser)
```

Reply:
This should be fixed in the latest commit!